### DISCUSSION TITLE
No close button on EditBehaviourDialog

### DISCUSSION THREAD

**Discussion by gautamv95** [4ian/GDevelop#2341] [Enhancement ideas (and their technical discussions)]
## Description
It would be great if the following dialog had a close button rather than having just apply!

![Screenshot_25](https://user-images.githubusercontent.com/45149542/77243837-afb20480-6c34-11ea-9568-21a19b626c92.png)

**Comment by 4ian** [comment]
Adding a cancel button is tricky because any change in the properties must result in calls to refactor the whole project (to rename the properties). Right now, any renaming is instantly applied, so we can't cancel. To have a cancel button, the dialog must be reworked so that Apply is launching the refactoring of the project to rename the properties that were renamed.

**Comment by arthuro555** [comment]
Huh is it that difficult? I would have thought this would only have been a matter of storing the values in the state and adding to the apply button a function that applies the states values to the behavior on click

**Comment by 4ian** [comment]
See my previous answer. To be more precise, the difficulty is about calls to `gd.WholeProjectRefactorer` in `_onBehaviorPropertyRenamed`. It has to be reworked so that all the renamed properties are triggering a refactoring of the whole project `gd.WholeProjectRefactorer.renameBehaviorProperty` when the dialog is closed using Apply. This means that you have to keep a state of renamed properties and give it to the apply method.

Currently, `EventsBasedBehaviorPropertiesEditor` is calling `onRenameProperty` as soon as a property is renamed - making "cancelling" impossible because the rest of the project was updated for the new property name.

Not super difficult, but "medium difficult" I would say - you have to understand properly how renaming is done.
